File: src/frontend/src/lib/custom-domain-record.ts
import type { CustomDomainRecordBnRegistrationState } from '@/declarations/backend/backend.did';

export const bnRegistrationIdFromBnRegistrationState = (
  state: CustomDomainRecordBnRegistrationState,
): string | undefined => {
  if ('pending' in state) {
    return state.pending.bn_registration_id;
  } else if ('registered' in state) {
    return state.registered.bn_registration_id;
  } else if ('failed' in state) {
    return state.failed.bn_registration_id;
  }
  return undefined;
};

export const bnRegistrationErrorMessageFromBnRegistrationState = (
  state: CustomDomainRecordBnRegistrationState,
): string | undefined => {
  if ('failed' in state) {
    return state.failed.error_message;
  }
  return undefined;
};

File: src/backend/impl/src/fixtures/custom_domain_record.rs
use rstest::*;

use crate::repositories::{
    CustomDomainRecord, CustomDomainRecordBnRegistrationState, TimestampFields,
};

#[fixture]
pub fn custom_domain_record_not_started() -> CustomDomainRecord {
    CustomDomainRecord {
        domain_name: "test.com".to_string(),
        bn_registration_state: CustomDomainRecordBnRegistrationState::NotStarted,
        timestamps: TimestampFields::new(),
    }
}

#[fixture]
pub fn custom_domain_record_pending() -> CustomDomainRecord {
    CustomDomainRecord {
        domain_name: "test.com".to_string(),
        bn_registration_state: CustomDomainRecordBnRegistrationState::Pending {
            bn_registration_id: "abcdefghijklmnopqrstuvwxyz".to_string(),
        },
        timestamps: TimestampFields::new(),
    }
}

#[fixture]
pub fn custom_domain_record_registered() -> CustomDomainRecord {
    CustomDomainRecord {
        domain_name: "test.com".to_string(),
        bn_registration_state: CustomDomainRecordBnRegistrationState::Registered {
            bn_registration_id: "abcdefghijklmnopqrstuvwxyz".to_string(),
        },
        timestamps: TimestampFields::new(),
    }
}

#[fixture]
pub fn custom_domain_record_failed() -> CustomDomainRecord {
    CustomDomainRecord {
        domain_name: "test.com".to_string(),
        bn_registration_state: CustomDomainRecordBnRegistrationState::Failed {
            bn_registration_id: "abcdefghijklmnopqrstuvwxyz".to_string(),
            error_message: "unknown error".to_string(),
        },
        timestamps: TimestampFields::new(),
    }
}
